### PR TITLE
Add chunked_filter (#344)

### PR DESCRIPTION
**Pull request by rafalkrupinski**

### CODE REVIEW

File: boltons/iterutils.py [hunk 1]
```diff
@@ -374,6 +374,30 @@ def chunked_iter(src, size, **kw):
     return
 
 
+def chunked_filter(iterable, predicate, chunk_size):
```
Comment (mahmoud) [inline boltons/iterutils.py:377]: Great to see a docstring here. Looking around at other functions, I noticed that most refer to the predicate as `key`. Do you mind updating this argument to be consistent?

File: boltons/iterutils.py [hunk 2]
```diff
@@ -374,6 +374,30 @@ def chunked_iter(src, size, **kw):
...
+    """A version of :func:`filter` which will call predicate with a chunk of the iterable.
```
Comment (mahmoud) [inline boltons/iterutils.py:378]: I think this is a great description, though it's a good idea to style arguments when referring to them (italicize or monospace).

**Review by mahmoud — CHANGES_REQUESTED** [review]
Overall looking good, would also be good to see an extra test or two for edge cases in `test_iterutils.py`. Thanks again for this!

File: boltons/iterutils.py (hunk 1 above)
Comment (rafalkrupinski) [reply boltons/iterutils.py:377]: I see various names. builtins.filter uses `iterable` and `function`. Actually it's [filter(function, iterable)](https://docs.python.org/3/library/functions.html#filter). Since it's a variation of it, how about using these names and order?

Comment (rafalkrupinski) [reply boltons/iterutils.py]: interestingly [itertools.filterfalse(predicate, iterable)](https://docs.python.org/3/library/itertools.html#itertools.filterfalse)

**Comment by rafalkrupinski** [comment]
builtins.filter allows None function, in which case an identity function is used. Here it doesn't make much sense from the use case perspective, but perhaps it would from the usability perspective and for consistency?

File: boltons/iterutils.py [hunk 3]
```diff
@@ -374,26 +374,50 @@ def chunked_iter(src, size, **kw):
     return
 
 
-def chunked_filter(iterable, predicate, chunk_size):
-    """A version of :func:`filter` which will call predicate with a chunk of the iterable.
+def chunked_filter(iterable, predicate, size):
```
Comment (mahmoud) [inline boltons/iterutils.py:377]: oops, based on the docstring, I think you meant to rename `predicate` here to `key`.

File: boltons/iterutils.py [hunk 4]
```diff
@@ -374,26 +374,50 @@ def chunked_iter(src, size, **kw):
...
     [0, 2, 4, 6, 8]
 
     In the above example the lambda function is called twice: once with values
     0-4 and then for 5-9.
 
     Args:
         iterable (Iterable): Items to filter
-        predicate (Callable): Predicate function
-        chunk_size (int): The maximum size of chunks that will be passed the
+        predicate (Callable): Bulk predicate function that accepts a list of items
```
Comment (mahmoud) [inline boltons/iterutils.py:388]: `predicate` -> `key` to match description above?

File: boltons/iterutils.py [hunk 5]
```diff
@@ -374,26 +374,50 @@ def chunked_iter(src, size, **kw):
...
+        raise TypeError('expected callable key')
+
+    def predicate_(src_):
+        allow_iter = predicate(src_)
+        if not is_iterable(allow_iter):
+            raise TypeError('expected an iterable from key(src)')
+
+        allow_list = list(allow_iter)
+        if len(allow_list) != len(src_):
+            raise ValueError('expected the iterable from key(src) has the same length as the passed chunk of items')
```
Comment (mahmoud) [inline boltons/iterutils.py:413]: "has the same" -> "to have the same"

I think it's great you're thinking about this. For exceptions, these days, I usually recommend the following format to maximize debuggability:

`raise ValueError("chunked_filter expected key func to return an iterable of length {src_len}, got {allow_list_len}")`.

Similar changes could be made to other exception messages above just by adding ", not {actually_received_type}"

**Review by mahmoud — CHANGES_REQUESTED** [review]
Good progress! I like all the edge case handling. I agree that there's no harm in allowing `key` to be `None`, except that then you may have to come up with a default for `size`, based on the ordering. I'm fine either way, so I'll let you make the call. :)

File: boltons/iterutils.py (hunk 3 above)
Comment (rafalkrupinski) [reply boltons/iterutils.py:377]: not sure, see previous comments

File: boltons/iterutils.py (hunk 1 above)
Comment (mahmoud) [reply boltons/iterutils.py]: Ah, I think the prime reference for this should be the [`sorted()`](https://docs.python.org/3/library/functions.html#sorted) function, where `key` is used. `min()` and `max()` also use `key`, I think. So, tldr. `key` should be preferred here. :)

Comment (rafalkrupinski) [reply boltons/iterutils.py]: > Ah, I think the prime reference for this should be the [`sorted()`](https://docs.python.org/3/library/functions.html#sorted) function, where `key` is used.

There's a good reason why sorted and min/max use `key` and `filter` & `filterfalse` don't. When you compare two objects, you either compare the value itself or its part, hence the `key` - attribute, only shorter. When you filter a collection, it can be an arbitrary criterium that's external to the object, so you use `function` (...to be called on an item) or more specifically, `predicate`.
Also, since it's an extension of `builtins.filter`, what's the reason for following naming pattern of any other function?
